### DISCUSSION TITLE
How to use all my CPU cores with Elysia?

### DISCUSSION THREAD

**Discussion by stefanos82** [elysiajs/elysia#416] [Q&A] [answered]
I have tried to find an example how to use all my CPU cores and couldn't find anything.

Can someone share their snippet if possible?

I'm interested in Elysia and would like to learn how to push its limits.

**Comment by loclv** [comment, marked as answer]
@stefanos82 

I think this is common question for running Bun/Node.js on across all CPUs available, without any code modifications.

You can read about how to do it with [pm2 - Cluster Mode](https://pm2.keymetrics.io/docs/usage/cluster-mode/).
And how to run [Bun as a daemon with PM2](https://bun.sh/guides/ecosystem/pm2).

Besides, you can use [Bun - workers](https://bun.sh/docs/api/workers)

<img width="1079" alt="Screenshot 2024-01-15 at 15 33 08" src="https://github.com/elysiajs/elysia/assets/6302717/17774d7f-9ad3-4b0c-95a4-3bf2d57e4a11">

**Comment by stefanos82** [comment]
Thank you @loclv :+1:

**Comment by stefanos82** [comment]
It seems I did not need to use the Node.js cluster after all; Bun supports [spawn](https://bun.sh/docs/api/spawn) which I took advantage of and produced the desired results I wanted from the very beginning.

Below is the code which you can place in a `spawn.ts` file and execute it via `bun spawn.ts`:

```ts
import os from 'node:os';

// @ts-ignore
const numCPUs = os.availableParallelism();
for (let i = 0; i < numCPUs; i++) {
    Bun.spawn(['bun', 'src/index.ts'], {
        stdio: ['inherit', 'inherit', 'inherit'],
        env: { ...process.env },
    });
}
```

With `wrk -c 100 -t 4 --latency http://127.0.0.1:3000/` I've managed to get:

```
Running 10s test @ http://127.0.0.1:3000/
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   697.66us    0.90ms  15.11ms   87.23%
    Req/Sec    52.88k     4.80k   64.45k    66.75%
  Latency Distribution
     50%  310.00us
     75%    0.95ms
     90%    1.83ms
     99%    4.08ms
  2108813 requests in 10.03s, 257.42MB read
Requests/sec: 210260.53
Transfer/sec:     25.67MB
```

What's interesting, without `spawn` the default single threaded `bun run src/index.ts` produces

```
Running 10s test @ http://127.0.0.1:3000/
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.87ms  152.01us   7.14ms   92.55%
    Req/Sec    28.95k     2.05k   58.95k    98.26%
  Latency Distribution
     50%  839.00us
     75%    0.87ms
     90%    0.92ms
     99%    1.38ms
  1158176 requests in 10.10s, 141.38MB read
Requests/sec: 114671.46
Transfer/sec:     14.00MB
```

which does not match the node.js cluster behavior, that is single-threaded result times the number of CPU cores available, in my case 4.

Is this an expected behavior?

**Reply by loclv** [reply]
@stefanos82 
Thank you for sharing! I appreciate this way a lot!
